### PR TITLE
CATROID-1139 Refactor controller and finder of ScriptFragment to kotlin

### PR DESCRIPTION
**Pull request by khushal1707**
Ticket: https://jira.catrob.at/browse/CATROID-1139
- Refactored ScriptController and ScriptFinder to kotlin
- Applied view-binding in ScriptFinder

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+        }
+        destinationSprite.scriptList.add(script)
+    }
+
+    @Throws(CloneNotSupportedException::class)
+    fun unpack(scriptToUnpack: Script, destinationSprite: Sprite) {
+        val script = scriptToUnpack.clone()
+        copyBroadcastMessages(script.scriptBrick)
+        for (brick in script.brickList) {
+            if (ProjectManager.getInstance().currentProject.isCastProject &&
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt:155]: please use injected projectManager instead.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+            }
+            copyBroadcastMessages(brick)
+        }
+        destinationSprite.scriptList.add(script)
+    }
+
+    private fun copyBroadcastMessages(brick: Brick): Boolean {
+        if (brick is BroadcastMessageBrick) {
+            val broadcastMessage = brick.broadcastMessage
+            return ProjectManager.getInstance().currentProject.broadcastMessageContainer.addBroadcastMessage(
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt:169]: same.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+    fun unpackForSprite(
+        scriptToUnpack: Script,
+        destinationProject: Project?,
+        destinationScene: Scene?,
+        destinationSprite: Sprite
+    ) {
+        val script = scriptToUnpack.clone()
+        for (brick in script.brickList) {
+            when {
+                ProjectManager.getInstance().currentProject.isCastProject && CastManager.unsupportedBricks.contains(
```
Comment (ndrnour) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt:186]: same.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt (hunk 1 above)
Comment (khushal1707) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt]: Done!

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt (hunk 2 above)
Comment (khushal1707) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt]: done

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt (hunk 3 above)
Comment (khushal1707) [reply catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt]: Done

**Comment by ndrnour** [comment]
the search script feature seems not to work anymore. please fix it. see attached screen recording.

**Comment by ndrnour** [comment]
https://user-images.githubusercontent.com/18437036/127777176-b375913f-6828-4262-81e5-e43b7ea98207.mp4

**Review by ndrnour — CHANGES_REQUESTED** [review]
the search script feature seems not to work anymore. please fix it. see attached screen recording.

**Comment by khushal1707** [comment]
> the search script feature seems not to work anymore. please fix it. see attached screen recording.

Fixed it! Working fine now.

**Review by ndrnour — APPROVED** [review]

File: catroid/src/main/java/org/catrobat/catroid/ui/ScriptFinder.kt [hunk 4]
```diff
@@ -0,0 +1,304 @@
...
+
+            override fun afterTextChanged(s: Editable) = Unit
+        }
+        binding.searchBar.addTextChangedListener(textWatcher)
+    }
+
+    companion object {
+        val TAG = ScriptFinder::class.java.simpleName
+        @Suppress("ComplexMethod", "TooGenericExceptionCaught")
+        fun searchBrickViews(v: View?, searchQuery: String?): Boolean {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/ScriptFinder.kt:97]: Please, simplify this.

File: catroid/src/main/java/org/catrobat/catroid/ui/ScriptFinder.kt [hunk 5]
```diff
@@ -0,0 +1,304 @@
...
+            ToastUtil.showError(
+                context,
+                context.getString(R.string.query_field_is_empty)
+            )
+        }
+    }
+
+    private fun findNext() {
+        if (searchResults?.isNotEmpty() == true) {
+            searchResultIndex = (searchResultIndex + 1) % searchResults!!.size
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/ScriptFinder.kt:149]: No `!!`

File: catroid/src/main/java/org/catrobat/catroid/ui/ScriptFinder.kt [hunk 6]
```diff
@@ -0,0 +1,304 @@
...
+            searchResults?.size?.let { it1 ->
+                onResultFoundListener?.onResultFound(
+                    it[0], it[1], it[2], it1,
+                    binding.sceneAndSpriteName
+                )
+            }
+        }
+    }
+
+    @Suppress("ComplexMethod")
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/ScriptFinder.kt:185]: Please, clean the code.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt [hunk 7]
```diff
@@ -0,0 +1,275 @@
...
+    @Throws(IOException::class, CloneNotSupportedException::class)
+    fun packForSprite(scriptToPack: Script, destinationSprite: Sprite) {
+        val script = scriptToPack.clone()
+        val scriptFlatBrickList: List<Brick> = ArrayList()
+        script.addToFlatList(scriptFlatBrickList)
+
+        for (brick in scriptFlatBrickList) {
+            when {
+                brick is SetLookBrick && brick.look != null ->
+                    brick.look = lookController.packForSprite(brick.look, destinationSprite)
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt:137]:
```suggestion
            when(brick) {
                is SetLookBrick, is WhenBackgroundChangesBrick ->
                    brick.look = brick.look?.let{lookController.packForSprite(it, destinationSprite)}
```
https://kotlinlang.org/docs/control-flow.html#when-expression
Improve all others too, please.

**Review by gPathpp — CHANGES_REQUESTED** [review]
https://medium.com/swlh/12-steps-to-convert-your-java-class-to-kotlin-the-right-way-9a718cfb498d

Have a close look at steps 4, 7, 8 and 12, please.

File: catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt [hunk 8]
```diff
@@ -0,0 +1,275 @@
...
+        return script
+    }
+
+    @Suppress("NestedBlockDepth")
+    @Throws(CloneNotSupportedException::class)
+    fun pack(groupName: String?, bricksToPack: List<Brick>?) {
+        val scriptsToPack: MutableList<Script> = ArrayList()
+        val userDefinedBrickListToPack: MutableList<UserDefinedBrick> = ArrayList()
+        bricksToPack?.let {
+            for (brick in bricksToPack) {
```
Comment (gPathpp) [inline catroid/src/main/java/org/catrobat/catroid/ui/recyclerview/controller/ScriptController.kt:111]: Use for each, please

**Pull request merged**